File: bin/gender/genderize.cpp
#include <vector>
#include <iostream>
#include <fstream>
#include "json.cpp"
#include "stringmap.h"

using namespace std;

const int MALE = 0;
const int FEMALE = 1;
const int UNDETERMINED = 2;

bool loadLinearSVCJSONFromFile(const string& filename, Json::Value& svcJSON)
{
    ifstream file;
    file.open(filename, ios::binary | ios::in);
    if (!file.is_open())
        return false;

    std::stringstream buffer;
    buffer << file.rdbuf();
    file.close();
    Json::Reader reader;
    return reader.parse(buffer.str(), svcJSON);
}

void toLowerCase(string &str)
{
    const int length = str.length();
    for(int i = 0; i < length; i++)
        str[i] = tolower(str[i]);
}

class GenderClassifier
{
public:
    GenderClassifier(const string& filename)
    {
        ngramCoefficients = NULL;
        nameGenderMap = NULL;
        Json::Value svcJSON;
        if (loadLinearSVCJSONFromFile(filename, svcJSON)) {
            // Load the intercept and threshold.
            uncertaintyThreshold = svcJSON["uncertainty_threshold"].asDouble();
            intercept = svcJSON["intercept"].asDouble();
            // Load the SVM ngram coefficients.
            Json::Value coefficientsJSON = svcJSON["coefficients"];
            ngramCoefficients = new ConstantSizeStringMap<float>(coefficientsJSON.size(), 0.0);
            for(Json::ValueIterator iter = coefficientsJSON.begin() ; iter != coefficientsJSON.end() ; iter++) {
                string ngram = iter.key().asString();
                ngramCoefficients->insert(ngram, coefficientsJSON[ngram].asDouble());
            }
            // Load the male and female names.
            Json::Value maleNamesJSON = svcJSON["male_names"];
            Json::Value femaleNamesJSON = svcJSON["female_names"];
            nameGenderMap = new ConstantSizeStringMap<int>(maleNamesJSON.size() + femaleNamesJSON.size(), UNDETERMINED);
            for(int index = 0; index < maleNamesJSON.size(); index++)
                nameGenderMap->insert(maleNamesJSON[index].asString(), MALE);

            for(int index = 0; index < femaleNamesJSON.size(); index++)
                nameGenderMap->insert(femaleNamesJSON[index].asString(), FEMALE);
        }
    }

    ~GenderClassifier()
    {
        if (ngramCoefficients) delete ngramCoefficients;
        if (nameGenderMap) delete nameGenderMap;
    }

    int predict(string name)
    {
        toLowerCase(name);
        int prediction = nameGenderMap->get(name);
        if (prediction != UNDETERMINED)
            return prediction;

        float distance = signedDistance(name);
        if (distance > uncertaintyThreshold)
            return FEMALE;

        else if (distance < -uncertaintyThreshold)
            return MALE;

        return UNDETERMINED;
    }

    bool successfullyInitialized()
    {
        return ngramCoefficients && nameGenderMap;
    }

private:
    float signedDistance(const string& name) {
        float dotproduct = 0;
        int norm = 0;
        string nameWithStartAndEndChars = "^" + name + "$";
        size_t nameLength = nameWithStartAndEndChars.size();
        // For 2-, 3- and 4-grams...
        for (int n = 2; n < 5; n++) {
            // For each consecutive chunk of characters...
            for (int cursor = 0; cursor <= nameLength - n; cursor++) {
                string gram = nameWithStartAndEndChars.substr(cursor, n);
                float coefficient = ngramCoefficients->get(gram);
                if (coefficient != 0.0) {
                    dotproduct += coefficient;
                    norm++;
                }
            }
        }
        return norm != 0 ? dotproduct / norm + intercept : 0;
    }

    ConstantSizeStringMap<float>* ngramCoefficients;
    ConstantSizeStringMap<int>* nameGenderMap;
    float intercept;
    float uncertaintyThreshold;
};

int main(int argc, char* argv[])
{
    GenderClassifier genderClassifier("bin/gender/linsvc.json");
    if (!genderClassifier.successfullyInitialized()) {
        cout << "{\"error\": \"the server failed to parse linear SVC data.\"}";
        return 1;
    }
    ConstantSizeStringMap<bool> inputNames(argc, false);
    cout << "{";
    for (int i = 1; i < argc; i++) {
        string name = string(argv[i]);
        if (inputNames.get(name))
            continue;

        if (i != 1)
            cout << ",";

        cout << "\"" << name << "\"" << ":" << genderClassifier.predict(name);
        inputNames.insert(name, true);
    }
    cout << "}";
    cout.flush();
    return 0;
}

File: bin/gender/genderize.py
import json
import sys

DEFAULT_UNCERTAINTY_THRESHOLD = 0.1
MALE_LABEL = 0
FEMALE_LABEL = 1
UNCERTAIN_LABEL = 2

def filename_with_extension(filename, extension):
    extension = "." + extension if extension[0] != "." else extension
    return filename + extension if filename[-len(extension):] != extension else filename

def load_json_as_object(filename):
    filename = filename_with_extension(filename, "json")
    jsonFile = open(filename, "r")
    dataAsString = jsonFile.read()
    jsonFile.close()
    return json.loads(dataAsString)

def ngrams_in_name(name):
    name = "^" + name + "$"
    grams = []
    for index, char in enumerate(name):
        for i in xrange(1, 4):
            if index < len(name) - i:
                grams.append(name[index:index + i + 1])
    return grams

class GenderClassifier(object):
    def __init__(self, filename):
        params = load_json_as_object(filename)
        self.maleNames = set(params["male_names"])
        self.femaleNames = set(params["female_names"])
        self.intercept = params["intercept"]
        self.coefficients = params["coefficients"]
        self.uncertaintyThreshold = params["uncertainty_threshold"]

    def _signed_distance(self, name):
        dotProduct, norm = 0, 0
        for ngram in ngrams_in_name(name):
            if ngram in self.coefficients:
                dotProduct += self.coefficients[ngram]
                norm += 1
        return (dotProduct / norm) + self.intercept if norm > 0 else 0

    def _label_from_signed_distance(self, distance):
        if abs(distance) < self.uncertaintyThreshold:
            return UNCERTAIN_LABEL
        return MALE_LABEL if distance <= 0 else FEMALE_LABEL

    def _label_from_data(self, name):
        if name in self.maleNames:
            return MALE_LABEL
        if name in self.femaleNames:
            return FEMALE_LABEL
        return UNCERTAIN_LABEL

    """
    Given a first name, returns 0, 1, or 2, indicating the gender of the
    name. 0 indicates male, 1 indicates female, and 2 indicates that the
    classifier was unable to reach any conclusion.
    """
    def gender(self, name):
        name = name.lower()
        genderFromData = self._label_from_data(name)
        if genderFromData is not UNCERTAIN_LABEL:
            return genderFromData
        return self._label_from_signed_distance(self._signed_distance(name))

if __name__ == "__main__":
    cls = GenderClassifier("bin/gender/linsvc.json")
    response = {}
    print json.dumps({ name: cls.gender(name) for name in sys.argv[1:] })

